Pattern Directory: Create endpoints to proxy api.w.org/patterns.

iandunn opened the pull request:
Fixes #26577

## Description

This adds API endpoints to browse and search for patterns. Similar to the Block Directory, the endpoints in Gutenberg proxy the api.wordpress.org endpoints, to prevent w.org from receiving the user's IP address.

* http://{{ hostname }}/wp-json/wp/v2/pattern-directory/
* http://{{ hostname }}/wp-json/wp/v2/pattern-directory/?category=2
* http://{{ hostname }}/wp-json/wp/v2/pattern-directory/search?term=image

## How has this been tested?

- [x] Manual testing with Insomnia
- [x] PHPUnit tests

## Types of changes

New feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

iandunn commented:
@TimothyBJacobs , heads up about this, in case you have any early feedback or thoughts.

iandunn commented:
The phpunit-watcher stuff is just cherry-picked from #27058, for convenience. I'll remove it before publishing the PR.

TimothyBJacobs commented on lib/class-wp-rest-pattern-directory-controller.php:
```diff
@@ -0,0 +1,209 @@
+
+			case 'query_patterns':
+				$prepared_pattern = (array) $raw_pattern->{'_embedded'}->self[0];
+				$prepared_pattern = array_intersect_key( $prepared_pattern, array_flip( self::SAFELISTED_FIELDS ) );
```
Instead of pulling from a separate constant, I'd look at the schema's properties.

iandunn replied:
👍🏻

iandunn replied:
Fixed in 6226e18

TimothyBJacobs commented:
What's the difference between `get_patterns` and `query_patterns`? Will the response format be different based on that? Ideally, it wouldn't. The search endpoint should have a uniform schema.

iandunn commented:
The "get" endpoint is for browsing, like https://api.wordpress.org/patterns/1.0/?action=get_patterns or https://api.wordpress.org/patterns/1.0/?action=get_patterns&category=2. 

The response format is different; the `get_patterns` endpoint on api.w.org are proxying a CPT's default endpoint, while `query_patterns` is proxying Core's `search` endpoint. There's also some workarounds due to https://core.trac.wordpress.org/ticket/49985 and https://core.trac.wordpress.org/ticket/49538.

The api.w.org code is at https://meta.trac.wordpress.org/browser/sites/trunk/api.wordpress.org/public_html/patterns/1.0/index.php?rev=10414

I imagined that `prepare_item_for_response()` would normalize the responses, though, so that searching and browsing would have the same schema.

TimothyBJacobs commented:
Ah, gotcha.

> I imagined that prepare_item_for_response() would normalize the responses, though, so that searching and browsing would have the same schema.

That'd be great!

iandunn commented:
The e2e test failure looks unrelated.

iandunn commented:
Should this go behind a feature flag or experimental namespace until the other parts of the pattern directory are implemented?

TimothyBJacobs commented:
> Should this go behind a feature flag or experimental namespace until the other parts of the pattern directory are implemented?

From a REST API perspective, I think it should have an experimental namespace. Whether it also needs to be behind a feature flag, I'm not sure. Maybe @youknowriad has thoughts about which features warrant that behavior.

TimothyBJacobs commented on lib/class-wp-rest-pattern-directory-controller.php:
```diff
@@ -0,0 +1,274 @@
+	}
+
+	/**
+	 * Checks whether a given request has permission to install and activate plugins.
```
doc block mentions plugins.

TimothyBJacobs commented on lib/class-wp-rest-pattern-directory-controller.php:
```diff
@@ -0,0 +1,274 @@
+	 * @return WP_Error|bool True if the request has permission, WP_Error object otherwise.
+	 */
+	public function get_items_permissions_check( $request ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedParameter
+		if ( current_user_can( 'edit_posts' ) ) {
```
We need to account for other post types that have a different capability set. See the themes controller for instance.

TimothyBJacobs commented on lib/class-wp-rest-pattern-directory-controller.php:
```diff
@@ -0,0 +1,274 @@
+
+		return new WP_Error(
+			'rest_pattern_directory_cannot_view',
+			__( 'Sorry, you are not allowed to browse the local block pattern directory. Please use https://api.wordpress.org/patterns/1.0/ for publicly-available pattern data.', 'gutenberg' ),
```
I think that including the URL might be a bit too technical? I don't think an end user would find that actionable. Maybe something to keep in the docs?

TimothyBJacobs commented on lib/class-wp-rest-pattern-directory-controller.php:
```diff
@@ -0,0 +1,274 @@
+	 */
+	public function prepare_item_for_response( $raw_pattern, $action ) {
+		switch ( $action ) {
+			default:
```
This is a very much a nit, but in most `switch` statements, the `default` is the last case. Could we maybe reverse these so `query_patterns` is first?

TimothyBJacobs commented on lib/class-wp-rest-pattern-directory-controller.php:
```diff
@@ -0,0 +1,274 @@
+		// Remove irrelevant properties.
+		$prepared_pattern = array_intersect_key( $prepared_pattern, $this->get_item_schema()['properties'] );
+
+		unset( $prepared_pattern['meta'] );
```
Why isn't that covered by the `array_intersect_key`?

TimothyBJacobs commented on lib/class-wp-rest-pattern-directory-controller.php:
```diff
@@ -0,0 +1,274 @@
+			'properties' => array(
+				'id'             => array(
+					'description' => __( 'The pattern ID.', 'gutenberg' ),
+					'type'        => 'int',
```
```suggestion
					'type'        => 'integer',
```

TimothyBJacobs commented on lib/class-wp-rest-pattern-directory-controller.php:
```diff
@@ -0,0 +1,274 @@
+
+				'category_ids'   => array(
+					'description' => __( "The pattern's category IDs.", 'gutenberg' ),
+					'type'        => 'array',
```
Let's give this an `items` definition.

TimothyBJacobs commented on lib/class-wp-rest-pattern-directory-controller.php:
```diff
@@ -0,0 +1,274 @@
+
+				'description'    => array(
+					'description' => __( 'A description of the pattern.', 'gutenberg' ),
+					'type'        => 'array',
```
Shouldn't this be a `string`.

TimothyBJacobs commented on phpunit/class-wp-rest-pattern-directory-controller-test.php:
```diff
@@ -0,0 +1,385 @@
+			default:
+			case 'get':
+				// Response from https://api.wordpress.org/patterns/1.0/?action=get_patterns.
+				$response = '[{"id":19,"title":{"rendered":"Two images side by side"},"content":{"rendered":"\n<figure class=\"wp-block-gallery alignwide columns-2 is-cropped\"><ul class=\"blocks-gallery-grid\"><li class=\"blocks-gallery-item\"><figure><img src=\"https:\/\/s.w.org\/images\/core\/5.5\/don-quixote-05.jpg\" alt=\"An old pencil drawing of Don Quixote and Sancho Panza sitting on their horses, by  ...
```
I think it might be better to move these long JSON definitions into a test fixture file.

TimothyBJacobs commented on lib/class-wp-rest-pattern-directory-controller.php:
```diff
@@ -0,0 +1,274 @@
+
+				'viewport_width' => array(
+					'description' => __( 'The preferred width of the viewport when previewing a pattern, in pixels.', 'gutenberg' ),
+					'type'        => 'array',
```
And this an `integer`?

TimothyBJacobs commented on lib/class-wp-rest-pattern-directory-controller.php:
```diff
@@ -0,0 +1,274 @@
+	}
+
+	/**
+	 * Retrieves the search params for the blocks collection.
```
Mentioning the blocks controller.

TimothyBJacobs commented on lib/class-wp-rest-pattern-directory-controller.php:
```diff
@@ -0,0 +1,274 @@
+		$query_params['context']['default'] = 'view';
+
+		/**
+		 * Filter collection parameters for the block directory controller.
```
Mentioning the block directory controller.

TimothyBJacobs reviewed (changes requested):
Some minor bits of feedback.

TimothyBJacobs commented:
Is there a bit more context of how the patterns structure is setup on .org? Looking just at `https://meta.trac.wordpress.org/browser/sites/trunk/api.wordpress.org/public_html/patterns/1.0/index.php`, I'm not sure why those need to be two different endpoints if the search endpoint is only looking in the `wporg-pattern` subtype. The purpose of the search controller is to allow for querying multiple/all subtypes at the same time.

iandunn commented:
Thanks for the review!

> I think it should have an experimental namespace

👍🏻 

That, and all the line-specific comments should be addressed in a45327f9f540356ce7b5f6ce402b9694a9704fd2.

>  I'm not sure why those need to be two different endpoints if the search endpoint is only looking in the wporg-pattern subtype. The purpose of the search controller is to allow for querying multiple/all subtypes at the same time.

Ah, that makes sense, I'll work on simplifying that. Thanks!

TimothyBJacobs commented on lib/class-wp-rest-pattern-directory-controller.php:
```diff
@@ -0,0 +1,281 @@
+	 * @return array Item schema data.
+	 */
+	public function get_item_schema() {
+		$this->schema = array(
```
The purpose of the `$schema` property is meant to cache the schema, so we should make sure to early return if it is specified. We need to call `add_additional_fields_schema` after the cache.

iandunn replied:
👍🏻 , fixed in 611dc4d

iandunn commented:
I updated api.w.org to no longer use `wp/v2/search` in https://meta.trac.wordpress.org/changeset/10572/, and 
2de7231c96f8cfecfdf25150129cec03ae752d5a updates this PR in light of that.

TimothyBJacobs commented:
Do we still need two separate endpoints in Gutenberg at this point?

iandunn commented:
🤔 , I don't think so. 

I'd assumed that it originally had two endpoints for consistency w/ the Block Directory, but that actually only has a `/search` endpoint. The Inserter only uses that to search for uninstalled blocks; it looks like the installed blocks are pre-loaded, so there isn't a need for a "browsing" endpoint.

I think the plan for patterns was to remove the hardcoded ones from Gutenberg, and load them all from w.org. Because of that, we'd want an endpoint for browsing (which can also be used for searching). @ryelle and @kellychoffman , does that sound right to you?

ryelle commented:
> the plan for patterns was to remove the hardcoded ones from Gutenberg, and load them all from w.org. Because of that, we'd want an endpoint for browsing (which can also be used for searching).

That's correct, we want to remove the patterns from Gutenberg/core and load those from w.org along with other new patterns, but plugins and themes will still be able to include them in code (so the local endpoint will need to take that into account). Then in Gutenberg, the user should be able to browse by category and search (and maybe filter in other ways we haven't thought of yet).

One API endpoint would likely be easier to manage from the Gutenberg side as well, so 👍  for that idea.

iandunn commented:
> One API endpoint would likely be easier to manage from the Gutenberg side as well, so 👍 for that idea.

👍🏻 , thanks!

> but plugins and themes will still be able to include them in code (so the local endpoint will need to take that into account). 

Do you mean that we'll need to leave the `wp-includes/block-patterns/*.php` files in Core for back-compat, because plugins/themes are already `include()`ing them? How do you see the endpoint taking that into account?

TimothyBJacobs commented:
I think @ryelle is referring to developers using `register_block_pattern` themselves and that those patterns should be returned from the REST API when the category/search terms match?

iandunn commented:
f68389bc31 consolodates the `/search` endpoint into `/`

face6639ef and 9078eb7913 are related. I think this is more appropriate than using `term`, but am curious what you think @TimothyBJacobs

ryelle commented:
What @TimothyBJacobs said - plugins and themes can still use `register_block_pattern`, and those locally-registered patterns should also be returned by the API.

TimothyBJacobs commented:
> [f68389b](https://github.com/WordPress/gutenberg/commit/f68389bc315e431990e14467c49d69c3e3656a08) consolodates the `/search` endpoint into `/`
> 
> [face663](https://github.com/WordPress/gutenberg/commit/face6639ef18c55ddc61b17146903b2b297ca7f2) and [9078eb7](https://github.com/WordPress/gutenberg/commit/9078eb79131b3747a649a0f7e6a7ba0ae8ab21c7) are related. I think this is more appropriate than using `term`, but am curious what you think @TimothyBJacobs

This seems good to me.

iandunn commented:
> those locally-registered patterns should also be returned by the API.

Ah, 👍🏻 

The schemas will need to match. 

Local patterns have `Title, content, description, viewportWidth, categories, keywords`, so we should probably add `keywords` to the w.org API. None of the Core patterns use it, but I'm assuming it's still desired.

Locally, `categories` is an array of slugs, rather than IDs, so we should probably return that instead. Although, local pattern would call `register_block_pattern_category()`, which asks for a `label` in addition to the slug. Maybe we don't need to handle that here, though, since that'd be handled by whatever code eventually "installs" the w.org pattern locally? That might depend on whether or not we have that "install" flow; IIRC that's not finalized (cc @kellychoffman, @DaisyOlsen)?

ryelle commented:
There really isn't a local install flow, the block code is just inserted when you click the pattern. (there has been some discussion about saving/favoriting patterns on wp.org, but we don't have that wp.org user connection on local WP sites, so the idea's been pushed off to later iterations)

You make a good point about categories though, we'll need some way of getting the categories list off wp.org + the local site to populate the category dropdown here:

![Screen Shot 2021-01-14 at 1 26 38 PM](https://user-images.githubusercontent.com/541093/104632756-2d377d00-566c-11eb-86a0-091b489b2d21.png)

On the wp.org side I've been assuming that's a static list of categories, but I don't think we want to make that assumption in the editor - that way we can add categories to wp.org without having to change Gutenberg.

iandunn commented:
>  we'll need some way of getting the categories list off wp.org + the local site to populate the category dropdown

👍🏻 , I think that'd be best scoped to a separate PR. I opened #28204 to track it.

TimothyBJacobs commented:
In light of that, I think we might want to go back to `pattern-directory/search` and have a  `pattern-directory/categories` endpoint that would return the list of categories.

iandunn commented:
> we might want to go back to `pattern-directory/search`

Do you mean renaming the current `/` to `/search`, or splitting them back into `/` for browsing and `/search` for searching?

TimothyBJacobs commented:
Renaming the current `/` to `/search`, or another name, but so we have "room" for a `/categories` endpoint.

iandunn commented:
👍🏻 , I'll do that, and normalize the local/remote schemas.

I don't think we should combine the local/remote patterns in this PR, though, since it opens a can of worms. I added #28248 for that. I think we should merge this once any final requests are done, and do the rest in separate PRs.

iandunn commented:
That latest batch of commits depends on some updates to w.org/patterns and api.w.org that I haven't finished & deployed yet.  the test fixtures are updates to reflect the upcoming changes, though.

i'll leave a note when the w.org endpoints are updated.

iandunn commented:
@TimothyBJacobs , is 5b92da59f52e91b847c2ea1312061472ea27855d the right way to do validation for the _response_? I dug through the docs, examples, etc, but just confused myself, and couldn't get `sanitize_callback` working. Is that only for the _request_?

TimothyBJacobs commented:
Yeah, all the built in validation/sanitization handling is for request data. We don't apply that level of sanitization in the block directory, but 🤷.

iandunn commented:
🤔 , it seems like a "better safe than sorry" situation to me. I wouldn't want Core to trust input any other remote APIs.

iandunn commented:
@TimothyBJacobs, Do you feel like this is ready to merge from an API perspective?

@youknowriad, would you like this behind a feature flag?

TimothyBJacobs commented on lib/class-wp-rest-pattern-directory-controller.php:
```diff
@@ -0,0 +1,262 @@
+	 */
+	public function get_items( $request ) {
+		$query_args   = array();
+		$category_ids = $request->get_param( 'category' );
+		$search_term  = $request->get_param( 'search' );
```
```suggestion
		$category_ids = $request['category'];
		$search_term  = $request['search']);
```

Preferred syntax for Core.

TimothyBJacobs commented on lib/class-wp-rest-pattern-directory-controller.php:
```diff
@@ -0,0 +1,262 @@
+		}
+
+		if ( $search_term ) {
+			$query_args['search'] = $search_term;
```
I think this need a `urlencode`, `add_query_args` expects encoded parameters.

iandunn replied:
Ah, great catch, thanks! Fixed in 30ad2af.

I'll see if I can make time to build a PR for https://github.com/WordPress/WordPress-Coding-Standards/issues/461 too

TimothyBJacobs commented on lib/class-wp-rest-pattern-directory-controller.php:
```diff
@@ -0,0 +1,262 @@
+
+			return $response;
+		}
+
```
Do we need to check that .org returned valid data and return an error if not? Looks like we do a `null` check in `plugins_api`.

iandunn replied:
Done in a1ec5d4 and 7ef3092

TimothyBJacobs commented on lib/class-wp-rest-pattern-directory-controller.php:
```diff
@@ -0,0 +1,262 @@
+	 *
+	 * @return array
+	 */
+	public function prepare_item_for_response( $raw_pattern, $request ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable -- Method must match signature of parent class.
```
This should return a `WP_REST_Response` object per the documented return type. `get_items` should then be adjusted to call `prepare_response_for_collection()`.

TimothyBJacobs commented on lib/class-wp-rest-pattern-directory-controller.php:
```diff
@@ -0,0 +1,262 @@
+			'$schema'    => 'http://json-schema.org/draft-04/schema#',
+			'title'      => 'pattern-directory-item',
+			'type'       => 'object',
+
```
```suggestion
```

TimothyBJacobs commented on lib/class-wp-rest-pattern-directory-controller.php:
```diff
@@ -0,0 +1,262 @@
+			'properties' => array(
+				'id'             => array(
+					'description' => __( 'The pattern ID.', 'gutenberg' ),
+					'required'    => true,
```
We should drop all these `required` declarations, they are used for documenting that a request parameter is `required`.

iandunn replied:
I added them because they're used in `rest_validate_value_from_schema()` (called from `assertPatternMatchesSchema()`). Is that not correct?

TimothyBJacobs replied:
I'd prefer manually adding them in the test so the published schema stays correct. `required` represents the request parameter is `required`, it doesn't say anything about the parameter being present in the response.

iandunn replied:
Ah, I see. Done in db207d157b

TimothyBJacobs commented on lib/class-wp-rest-pattern-directory-controller.php:
```diff
@@ -0,0 +1,262 @@
+					'required'    => true,
+					'type'        => 'integer',
+					'minimum'     => 1,
+					'context'     => array( 'view' ),
```
At least some of these should have an `embed` set so if a link to this controller is embedded, it returns fields. Since we don't have a single item endpoint, though we might want to consider adding one in the future, we should add `embed` to all these fields.

iandunn replied:
👍🏻 , added in 6fe5d9e42e

TimothyBJacobs commented on lib/class-wp-rest-pattern-directory-controller.php:
```diff
@@ -0,0 +1,262 @@
+			'keywords'       => array_map( 'sanitize_title', $raw_pattern->keyword_slugs ),
+			'description'    => sanitize_text_field( $raw_pattern->meta->wpop_description ),
+			'viewport_width' => absint( $raw_pattern->meta->wpop_viewport_width ),
+		);
```
We should also have an `add_additional_fields_to_object` call to match the `add_additional_fields_schema`.

iandunn replied:
added in a781ac66f0a2314e3392a36f93555e63ab121159

TimothyBJacobs commented on lib/class-wp-rest-pattern-directory-controller.php:
```diff
@@ -0,0 +1,262 @@
+	 * @return array Collection parameters.
+	 */
+	public function get_collection_params() {
+		$query_params = parent::get_collection_params();
```
If we don't support paginating, we should `unset` the `page` and `per_page` fields.

iandunn replied:
🤔  I'm not sure if we will or not, but there's no harm in disabling now. We can always enable if we discover we do need it.

iandunn replied:
removed in 447d228a4f

TimothyBJacobs commented on lib/class-wp-rest-pattern-directory-controller.php:
```diff
@@ -0,0 +1,262 @@
+			'description'       => __( 'Limit results to those matching a category ID.', 'gutenberg' ),
+			'type'              => 'integer',
+			'minimum'           => 1,
+			'sanitize_callback' => 'sanitize_text_field',
+			'validate_callback' => 'rest_validate_request_arg',
```
```suggestion
```

This will be validated by the schema.

TimothyBJacobs commented on phpunit/class-wp-rest-pattern-directory-controller-test.php:
```diff
@@ -0,0 +1,365 @@
+		);
+	}
+
+	public static function wpTearDownAfterClass() {
+		self::delete_user( self::$contributor_id );
+	}
+
```
```suggestion
```

This isn't necessary. It'll be dropped in the base test class with a call to `_delete_all_data`.

TimothyBJacobs commented on phpunit/class-wp-rest-pattern-directory-controller-test.php:
```diff
@@ -0,0 +1,365 @@
+		$schema     = ( new WP_REST_Pattern_Directory_Controller() )->get_item_schema();
+		$pattern_id = isset( $pattern['id'] ) ? $pattern['id'] : 'missing';
+
+		self::assertTrue(
```
```suggestion
		$this->assertTrue(
```

iandunn replied:
Fixed in c3e2847 and d118ddf

TimothyBJacobs reviewed (changes requested):
Left some small bits of feedback.

youknowriad commented:
> @youknowriad, would you like this behind a feature flag?

Since this is a PHP change that is only backported explicitly into Core. I don't think it should necessarily be behind a feature flag.  We can include in Core when we think it's ready or when it's being used by the frontend.

iandunn commented:
Thanks, that sounds good 👍🏻

TimothyBJacobs commented on lib/class-wp-rest-pattern-directory-controller.php:
```diff
@@ -0,0 +1,292 @@
+					__( 'An unexpected error occurred. Something may be wrong with WordPress.org or this server&#8217;s configuration. If you continue to have problems, please try the <a href="%s">support forums</a>.' ),
+					__( 'https://wordpress.org/support/forums/' )
+				),
+				wp_remote_retrieve_body( $wporg_response )
```
```suggestion
				array(
					'status' => 500,
					'response' => wp_remote_retrieve_body( $wporg_response ),
				);
```

Let's give this an explicit `500` status, and nest the raw body in a keyed field.

TimothyBJacobs reviewed (approved):
Looks great to me!

iandunn committed:
Pattern Directory: Add API endpoints for browsing and searching.

Similar to the Block Directory, the endpoints in Gutenberg proxy the api.wordpress.org endpoints, to prevent w.org from receiving the user's IP address. This will also be necessary in the future to combine locally-registered patterns with those hosted in the wordpress.org directory.

Fixes #26577

iandunn commented:
👍🏻 thanks for the review!

I squashed all the WIP commits down, and rebased against `master`. Is there anything else needed to merge?

TimothyBJacobs commented:
Looks great, merged.

iandunn commented:
Thanks!

mtias commented:
Awesome to see this progress :)
